### PR TITLE
feat(scheduler): add swarm scheduler support to deis controller

### PR DESCRIPTION
**Pull request by smothiki**
Install [deis](http://docs.deis.io/en/latest/) platform .
Follow the instructions to launch [controller]( http://docs.deis.io/en/latest/contributing/hacking/?highlight=hacking%20deis)
Deploy deis-swarm-node ```deisctl install swarm-node && deisctl start swarm-node```
Deploy deis-swarm-manager ```deisctl install swarm-manager && deisctl start swarm-manager```

set the scheduler config ```deisctl config controller set schedulerModule=swarm```

Swarm is in a very early stage of development . Doesn't support fail-over of swarm nodes and fail-over of containers .
Swarm follows bin packing scheduling strategy . We are working on providing a minimal HA strategy for App containers by using affinity flags .

### CODE REVIEW

File: builder/image/bin/boot [hunk 1]
```diff
@@ -54,7 +54,7 @@ CONFD_PID=$!
 test -e /var/run/docker.sock && rm -f /var/run/docker.sock
 
 # spawn a docker daemon to run builds
-docker -d --bip=172.19.42.1/16 --insecure-registry 10.0.0.0/8 --insecure-registry 172.16.0.0/12 --insecure-registry 192.168.0.0/16 --insecure-registry 100.64.0.0/10 &
+docker -d --bip=172.19.42.1/16 --insecure-registry 10.0.0.0/8 --insecure-registry 172.16.0.0/12 --insecure-registry 192.168.0.0/16 --insecure-registry 100.64.0.0/10 --storage-driver=overlay &
```
Comment (bacongobbler) [inline builder/image/bin/boot]: is this necessary for swarm? I thought that the daemon would select the best driver

File: contrib/coreos/user-data.example [hunk 2]
```diff
@@ -90,7 +106,7 @@ write_files:
     content: |
         [Service]
         EnvironmentFile=/etc/environment_proxy
-        Environment="DOCKER_OPTS=--insecure-registry 10.0.0.0/8 --insecure-registry 172.16.0.0/12 --insecure-registry 192.168.0.0/16 --insecure-registry 100.64.0.0/10"
+        Environment="DOCKER_OPTS=--insecure-registry 10.0.0.0/8 --insecure-registry 172.16.0.0/12 --insecure-registry 192.168.0.0/16 --insecure-registry 100.64.0.0/10 --insecure-registry registry.smothiki.io:5000"
```
Comment (bacongobbler) [inline contrib/coreos/user-data.example]: could you please remove your dev registry? thanks!

File: controller/templates/confd_settings.py [hunk 3]
```diff
@@ -10,6 +10,9 @@
 except:
     SCHEDULER_OPTIONS = {}
 
+# scheduler swarm manager host
+SWARM_HOST = """{{ or (.deis_scheduler_swarm_host) "" }}"""
```
Comment (bacongobbler) [inline controller/templates/confd_settings.py]: I believe we should be using the existing config option SCHEDULER_TARGET for this

File: builder/image/bin/boot (hunk 1 above)
Comment (mboersma) [inline builder/image/bin/boot]: Isn't overlay still experimental? I think you have to ask for it explicitly until it graduates to supported.

File: deisctl/units/deis-swarm-manager.service [hunk 4]
```diff
@@ -0,0 +1,15 @@
+[Unit]
+Description=deis-cache
+
+[Service]
+EnvironmentFile=/etc/environment
+TimeoutStartSec=20m
+ExecStartPre=/bin/sh -c "docker history smothiki/swarm >/dev/null || docker pull smothiki/swarm"
```
Comment (bacongobbler) [inline deisctl/units/deis-swarm-manager.service]: Is there something necessary in your custom image that docker's `swarm` image doesn't provide?

**Comment by bacongobbler** [comment]
From first impressions, looks like a nice implementation on bootstrapping swarm on top of Deis! It could use some clean-up, but it's a great PoC. Great work @smothiki :)

In regards to `smothiki/swarm`, it's just running `swarm join` on $HOST, then publishing itself to etcd. Those could be rewritten into a single ExecStart command, which would remove the need for our own custom swarm image. We could then run the native swarm image which IMO would be the better option here (less maintenance on our end).

File: builder/image/bin/boot (hunk 1 above)
Comment (smothiki) [inline builder/image/bin/boot]: docker 1.5.0 on core os uses overlay fs . We have updated docker version inside builder and started docker with overlay fs
```
$ docker info
Containers: 0
Images: 0
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Kernel Version: 3.18.6
Operating System: CoreOS 607.0.0
CPUs: 1
Total Memory: 1.959 GiB
Name: deis-03
ID: LCIO:LLZB:VOM3:T44O:HM5Y:6Z6Z:7UQT:V2UK:LNAI:HYTK:RFRZ:MOMR
```

File: deisctl/units/deis-swarm-manager.service (hunk 4 above)
Comment (smothiki) [inline deisctl/units/deis-swarm-manager.service]: We need to figure out etcd master to bootstrap swarm nodes and manager . The custom image has deis wrapper for swarm which starts swarm nodes and swarm managers and finds ETCD master. Custom image is builds swarm binary from my fork where I have soft affinity implementations and little experiments with health checks

**Comment by bacongobbler** [comment]
I talked to @smothiki offline on my last comment so we're good now on that front.

My question is now a bit more philosophical... If we're using fleet to bootstrap the application deployment process, shouldn't we be cutting fleet out entirely? I'm talking about:

 - deploying a CoreOS cluster with just etcd and swarm, no fleet
 - deisctl deploys the control plane via swarm

If I'm already going down the route of swarm, I'd imagine that I'd want to go all-out on swarm rather than relying on 2 separate schedulers for the same platform.

Is this something which we'd like to address in the future if we were to support swarm, or am I completely out of whack here?

**Comment by smothiki** [comment]
The main advantage using Fleet is systemd units and control plane is kind of cluster agnostic , which makes maintenance easy for control plane . The same thing can be acheived using swarm but we have to write out own mechanism for fail overs , Exec start pre , Exec Start Post etc .. I don't see any need to move control plane away from fleet in the near future .

**Comment by bacongobbler** [comment]
> The same thing can be achieved using swarm but we have to write our own mechanism for fail overs , Exec start pre , Exec Start Post etc

The `ExecStartPre`/`ExecStartPost` is all crud from explicitly trying to "orchestrate" docker with fleet. Fleet's a distributed init system... Swarm is a distributed docker orchestration tool. `docker run --restart=always -p 8000:8000 --name deis-controller  -e HOST=$(ip route show 0.0.0.0/0 | grep -Eo 'via \S+' | awk '{ print \$2 }') deis/controller:v1.4.1` accomplishes everything which the fleet units currently do today, *including* failover. The only thing that is truly missing is host fail-over which is coming Soon(:tm:).

Personally I think it's worth looking into (as well as docker machine) at some point in the future but I agree that it's not necessary in the near future.

/my2cents

File: builder/image/bin/boot [hunk 5]
```diff
@@ -66,16 +66,16 @@ DOCKER_PID=$!
...
 # HACK: load progrium/cedarish tarball for faster boot times
 # see https://github.com/deis/deis/issues/1027
 if ! docker history progrium/cedarish >/dev/null 2>/dev/null ; then
-    echo "Loading cedarish..."
-    cat /progrium_cedarish.tar.gz | docker import - progrium/cedarish
+		echo "Loading cedarish..."
+		cat /progrium_cedarish.tar.gz | docker import - progrium/cedarish
 else
-    echo "Cedarish already loaded"
+		echo "Cedarish already loaded"
```
Comment (mboersma) [inline builder/image/bin/boot]: It looks like the changes to this file are all whitespace-related, and inconsistent before and after... Can we omit this?

File: contrib/coreos/user-data.example [hunk 6]
```diff
@@ -113,7 +129,7 @@ write_files:
     content: |
         [Service]
         EnvironmentFile=/etc/environment_proxy
-        Environment="DOCKER_OPTS=--insecure-registry 10.0.0.0/8 --insecure-registry 172.16.0.0/12 --insecure-registry 192.168.0.0/16 --insecure-registry 100.64.0.0/10"
+        Environment="DOCKER_OPTS=--insecure-registry 10.0.0.0/8 --insecure-registry 172.16.0.0/12 --insecure-registry 192.168.0.0/16 --insecure-registry 100.64.0.0/10 "
```
Comment (mboersma) [inline contrib/coreos/user-data.example]: The trailing space inside the quotation marks should be removed.

File: controller/requirements.txt [hunk 7]
```diff
@@ -10,8 +10,8 @@ django-fsm==2.2.0
 django-guardian==1.2.5
 django-json-field==0.5.7
 djangorestframework==3.0.5
-docker-py==0.7.2
 gunicorn==19.3.0
+docker-py==1.0.0
```
Comment (mboersma) [inline controller/requirements.txt]: [docker-py 1.1.0](https://pypi.python.org/pypi/docker-py/1.1.0) is out now, just FYI.

File: deisctl/units/deis-swarm-manager.service (hunk 4 above)
Comment (mboersma) [inline deisctl/units/deis-swarm-manager.service]: We should make that `deis/swam` then, or something more official. :office:

File: controller/templates/confd_settings.py [hunk 8]
```diff
@@ -10,6 +10,11 @@
 except:
     SCHEDULER_OPTIONS = {}
 
+# scheduler swarm manager host
+#SWARM_HOST = '{{ if exists "/deis/scheduler/swarm/host" }}{{ getv "/deis/scheduler/swarm/host" }}{{ else }}127.0.0.1{{ end }}'
+#SWARM_HOST = '{{ or (.deis_scheduler_swarm_host) "127.0.0.1" }}'
```
Comment (mboersma) [inline controller/templates/confd_settings.py]: Are these meant to be commented out?

File: contrib/coreos/user-data.example [hunk 9]
```diff
@@ -34,6 +34,22 @@ coreos:
...
+
+      [Socket]
+      ListenStream=2375
+      BindIPv6Only=both
+      Service=docker.service
+
+      [Install]
+      WantedBy=sockets.target
+
+
```
Comment (mboersma) [inline contrib/coreos/user-data.example]: Let's remove these two blank lines. We don't seem to do that elsewhere in this file.

File: builder/image/bin/boot (hunk 5 above)
Comment (bacongobbler) [inline builder/image/bin/boot]: IIRC we don't have anything in place but I use 4 spaces per indentation block and @carmstrong follows 2 spaces per intendation block. Only think that should use tabs are stuff written in Go

Comment (mboersma) [inline builder/image/bin/boot]: Well even if we don't have a clear standard for shell indentation, this kind of change just makes it harder to review the overall diff and should be omitted IMHO.

Comment (bacongobbler) [inline builder/image/bin/boot]: :+1:

File: contrib/coreos/user-data.example (hunk 9 above)
Comment (smothiki) [inline contrib/coreos/user-data.example]: Ok

**Comment by mboersma** [comment]
This needs a rebase now.

**Comment by smothiki** [comment]
I'll update the PR with re-based code

**Comment by carmstrong** [comment]
@smothiki Do you think this will be ready to release in preview mode for v1.6?

**Comment by smothiki** [comment]
Yes , I'll confirm by next week Wednesday . I'm so close I spent enough
time thinking about failover mechanism .

On Thursday, April 16, 2015, Chris Armstrong <notifications@github.com>
wrote:

> @smothiki <https://github.com/smothiki> Do you think this will be ready
> to release in preview mode for v1.6?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/deis/deis/pull/3134#issuecomment-93778029>.
>


-- 
Thanks
Sivaram Mothiki

File: controller/templates/confd_settings.py [hunk 10]
```diff
@@ -10,6 +10,11 @@
 except:
     SCHEDULER_OPTIONS = {}
 
+# scheduler swarm manager host
+SWARM_HOST = '{{ if exists "/deis/scheduler/swarm/host" }}{{ getv "/deis/scheduler/swarm/host" }}{{ else }}127.0.0.1{{ end }}'
+#SWARM_HOST = '{{ or (.deis_scheduler_swarm_host) "127.0.0.1" }}'
```
Comment (gabrtv) [inline controller/templates/confd_settings.py]: Let's get rid of this comment.

File: swarm/swarm.go [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+func getleaderHost() string {
+	var response []Response
+	var host string
+	client := &http.Client{}
+	resp, _ := client.Get("http://" + os.Getenv("HOST") + ":7001/v2/admin/machines")
+	defer resp.Body.Close()
+	body, _ := ioutil.ReadAll(resp.Body)
+	// body, _ := ioutil.ReadFile("test.json")
```
Comment (carmstrong) [inline swarm/swarm.go]: Let's clean up all of these comments.

File: deisctl/units/deis-swarm-node.service [hunk 12]
```diff
@@ -0,0 +1,18 @@
+[Unit]
+Description=deis-cache
+
+[Service]
+EnvironmentFile=/etc/environment
+TimeoutStartSec=20m
+ExecStartPre=/bin/sh -c "docker history smothiki/swarm  >/dev/null || docker pull smothiki/swarm "
```
Comment (gabrtv) [inline deisctl/units/deis-swarm-node.service]: Need to change this to `deis/swarm:v<version>` also.

File: contrib/coreos/user-data.example (hunk 6 above)
Comment (mboersma) [inline contrib/coreos/user-data.example]: ping.

File: swarm/swarm.go (hunk 11 above)
Comment (gabrtv) [inline swarm/swarm.go]: Let's clean up comments in this file.

**Comment by mboersma** [comment]
Once existing comments are addressed and this settles down, please squash the 24 commits together into logical units. Commits like b0737cd9f4b11b659dcfaa5f1d10a9d448424178 shouldn't be listed, for example.

File: swarm/test.json [hunk 13]
```diff
@@ -0,0 +1,20 @@
+[
```
Comment (gabrtv) [inline swarm/test.json]: Let's remove this until we need it.

File: tests/utils/itutils.go [hunk 14]
```diff
@@ -168,12 +168,22 @@ func CurlWithFail(t *testing.T, url string, failFlag bool, expect string) {
 			if strings.Contains(string(err.Error()), expect) {
 				fmt.Println("(Error expected...ok) " + expect)
 			} else {
+				if expect == "503" {
```
Comment (gabrtv) [inline tests/utils/itutils.go]: This should be a separate PR.

File: builder/image/bin/boot [hunk 15]
```diff
@@ -66,16 +66,16 @@ DOCKER_PID=$!
 
 # wait for docker to start
 while [[ ! -e /var/run/docker.sock ]]; do
-  sleep 1
+	sleep 1
```
Comment (mboersma) [inline builder/image/bin/boot]: All the changes to this file appear to be whitespace-related, but we don't have a standard for shell script spacing. Could you revert the changes to builder/image/bin/boot? There is enough good stuff to review here already. :eyeglasses:

File: controller/scheduler/swarm.py [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+                                         environment=[affinity])
+        self.docker_cli.stop(name)
+
+    def start(self, name):
+        """
+        Start a container
+        """
+        self.docker_cli.start(name, publish_all_ports=True)
+
+        return
```
Comment (mboersma) [inline controller/scheduler/swarm.py]: Empty `return` statements aren't necessary in python methods--let's remove all these.

File: controller/scheduler/swarm.py [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+            if self.docker_cli.inspect_container(name)["State"]["Running"]:
+                return JobState.up
+            else:
+                return JobState.created
+        except Exception:
+            return JobState.destroyed
+
+    def state(self, name):
+        try:
+            # NOTE (bacongobbler): this call to ._get_unit() acts as a pre-emptive check to
```
Comment (mboersma) [inline controller/scheduler/swarm.py]: This comment looks like a copy-and-paste--it should be updated to match the local code.

File: controller/scheduler/swarm.py [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+            # determine if the job no longer exists (will raise a RuntimeError on 404)
+            for _ in range(30):
+                return self._get_container_state(name)
+                time.sleep(1)
+            # FIXME (bacongobbler): when fleet loads a job, sometimes it'll automatically start and
```
Comment (mboersma) [inline controller/scheduler/swarm.py]: This is a fleet-related comment that should probably be removed.

File: controller/scheduler/swarm.py [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+            # stop the container, which in our case will return as 'failed', even though
+            # the container is perfectly fine.
+        except KeyError:
+            # failed retrieving a proper response from the fleet API
```
Comment (mboersma) [inline controller/scheduler/swarm.py]: This comment refers to fleet.

File: controller/scheduler/swarm.py [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+            return JobState.error
+        except RuntimeError:
+            # failed to retrieve a response from the fleet API,
```
Comment (mboersma) [inline controller/scheduler/swarm.py]: This comment refers to fleet.

File: controller/templates/confd_settings.py [hunk 21]
```diff
@@ -10,6 +10,9 @@
 except:
     SCHEDULER_OPTIONS = {}
 
+# scheduler swarm manager host
+SWARM_HOST = '{{ if exists "/deis/scheduler/swarm/host" }}{{ getv "/deis/scheduler/swarm/host" }}{{ else }}127.0.0.1{{ end }}'
+
 # base64-encoded SSH private key to facilitate current version of "deis run"
```
Comment (smothiki) [inline controller/templates/confd_settings.py]: @gabrtv @carmstrong I want to keep the separate key spaces in the same PR . Yesterday I forgot to mention the two use cases where we require separate key spaces for schedulers.
##### User Forgot to set the config for schedulerModule
1. user ran ```deisctl install <scheduler> && deisctl start <scheduler>``` .
2. The scheduler will keep publishing the ```/deis/controller/schedulerTarget``` to scheduler endpoint. But still uses fleet with the endpoint of the new scheduler 

##### User Forgot to stop the scheduler
1. if a user forgots to stop scheduler and set the config for schedulerModule to a different scheduler 
2. Both the schedulers try to set ```deis/controller/schedulerTarget``` and confd keeps updating controller confd templates accordingly.

 I myself observed these errors, Yesterday I could recollect them

Comment (carmstrong) [inline controller/templates/confd_settings.py]: I'm fine with the separate keyspace.

**Comment by mboersma** [comment]
I have been testing this branch, and the Swarm Scheduler is working well. I have some docs to add and there are a couple minor issues we will try to fix today, but they are not showstoppers. This looks good so far. :bee:

File: deisctl/cmd/swarm.go [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+	go printState(outchan, errchan, 500*time.Millisecond)
+	outchan <- utils.DeisIfy("Starting Swarm...")
+	outchan <- fmt.Sprintf("swarm nodes...")
+	b.Start([]string{"swarm-node"}, &wg, outchan, errchan)
+	wg.Wait()
+	outchan <- fmt.Sprintf("swarm manager...")
+	b.Start([]string{"swarm-manager"}, &wg, outchan, errchan)
+	wg.Wait()
+	fmt.Println("Done.")
+	fmt.Println("Please run deisctl config controller set schedulerModule=swarm to change scheduler to swarm")
```
Comment (mboersma) [inline deisctl/cmd/swarm.go]: For consistency with other messages in `deisctl`, could you change this string to:
```
"Please run `deisctl config controller set schedulerModule=swarm` to use the swarm scheduler."
```

File: deisctl/cmd/platfrom.go [hunk 23]
```diff
@@ -0,0 +1,102 @@
+package cmd
```
Comment (mboersma) [inline deisctl/cmd/platfrom.go]: Typo: this file should be renamed to `platform.go`.

**Comment by mboersma** [comment]
This is ready for hopefully final review. I've tested it and added docs. Note that we may want to fix #3605 for v1.6 as well, since it can be triggered more easily with the Swarm Scheduler.

**Comment by mboersma** [comment]
Closing this in favor of #3612 after the `git` gymnastics became too difficult to maintain here.

**Pull request closed without merging**